### ISSUE TITLE
Gamescope doesn't quit when game quits

### ISSUE THREAD

**Issue by shindaseishin** [Plagman/gamescope#777]
OS: KUbuntu 22.10
Kernels tested: stock ubuntu 5.19.0 and xanmod 6.1
GPU: AMD RX 570
Mesa version: 22.2.5

Tested on both Plasma X11 and Plasma Wayland.

When I play a game via steam using gamescope everything works correctly until I quit the game.

When I do, the game ends, music stops playing, etc. but I am not dropped out to desktop. Steam shows the game is still running. I can bring up the process manager and see that the game process is no longer running but an instance of gamescope will still be running. Killing it will make the game disappear and the desktop to return. I can also kill the gamescope process by pressing stop in the steam client. It happens both full screen and windowed.

I have tested with Pillars of Eternity: Deadfire (Linux native), Valheim (native), Aragami 2 (proton), and Portal (native). All exhibit the same behaviour.

I also tried Stardew Valley (with SMAPI installed) but that exited correctly. Not sure why that worked when the others did not.

Compiled git master branch (fully up to date as of about an hour ago).

Happy to provide any additional details if they would help.

**Comment by iggut** [comment]
Same here in Ubuntu 22.10 Gnome wayland.
Compiled it today and tried Looking-Glass-Client for my Windows 11 VM with gamescope and it works very well.
When shutting down windows it closes looking-glass, but the terminal that was used to launch gamescope/looking-glass needs a Ctrl+C to stop the process.

**Comment by patlefort** [comment]
A little debugging show that it is stuck on https://github.com/Plagman/gamescope/blob/2a09fbcb695102e9a990ed7cf1fdc350397a58f1/src/wlserver.cpp#L1211

Adding a timeout fix the issue, however I don't know why it is stuck there.

**Comment by Faugus** [comment]
I'm having this problem with Life is Strange: Before the Storm.
If I Alt+F4, Gamescope closes.
But if I quit the game from the main menu, Gamescope won't close.


> A little debugging show that it is stuck on
> 
> https://github.com/Plagman/gamescope/blob/2a09fbcb695102e9a990ed7cf1fdc350397a58f1/src/wlserver.cpp#L1211
> 
> Adding a timeout fix the issue, however I don't know why it is stuck there.

How did you add this timeout?

**Comment by patlefort** [comment]
> How did you add this timeout?

By changing -1 in that line to a value in milliseconds, like 1000.

**Comment by shindaseishin** [comment]
I tried adding the timeout as @patlefort suggested and it worked. Thank you!

**Comment by Faugus** [comment]
> > How did you add this timeout?
> 
> By changing -1 in that line to a value in milliseconds, like 1000.

Is there a way to add this timeout to a already compiled build? Cause I'm using the Gamescope from Arch's repository.

**Comment by shindaseishin** [comment]
No. That value is hard coded into the program and can't be changed at run time, only compile time. Maybe someone could make a pull request with the change in it to get it into the main code base which would get in to the arch version eventually, but it is probably better to figure out why the original line is failing.

**Comment by DocMAX** [comment]
Shouldn't `gamescope -- sleep 3` close gamescope after 3 sec?

**Comment by shindaseishin** [comment]
Nope. It doesn't close after 3 seconds. I recompiled gamescope to current master without the fix @patlefort suggested and ran the command @DocMAX  posted. With the process monitor open I can see gamescope start and then a moment later sleep starts. Three seconds later sleep quits but gamescope keeps going. I have to manually kill it.

Just for testing purposes, I reran a few games with the recompiled version, Valheim still exhibits the issue, but Cyberpunk 2077 does not.

**Comment by DocMAX** [comment]
So is this a bug or works by design?

**Comment by shindaseishin** [comment]
It is a bug. gamescope doesn't quit. It stays resident in memory after it is supposed to have ended and continues to display a screen when it is not supposed to.

**Comment by mlgomez** [comment]
> A little debugging show that it is stuck on
> 
> https://github.com/Plagman/gamescope/blob/2a09fbcb695102e9a990ed7cf1fdc350397a58f1/src/wlserver.cpp#L1211
> 
> Adding a timeout fix the issue, however I don't know why it is stuck there.

This seems like something that should be in a pull request.

**Comment by awsms** [comment]
I noticed that killing gamescope won't kill the invoked program either.... is it expected behaviour, or another bug?

**Comment by EoceneMiacid** [comment]
This is really quite a severe bug.

It means I can start a Steam game in gamescope just fine, but I can't launch another one without rebooting the PC, even after explicitly killing the 'gamescope' process, restarting Steam, etc.

**Comment by denisstrizhkin** [comment]
I encounter the same problem with flatpak:
| ID | Version | Branch |
|-|-|-|
|com.valvesoftware.Steam.CompatibilityTool.Proton-GE | 8.14-1 | stable |
|org.freedesktop.Platform.VulkanLayer.gamescope|3.12.5|23.08|
|com.valvesoftware.Steam | 1.0.0.78 | stable |

My desktop environment is GNOME Wayland. When I launch Stardew Valley it works as expected and if I quit from game menu it closes just fine too. But if I close the window itself then the game closes but gamescope just hangs around. Even if I use `sudo pkill -9 -f gamescope` the game shows as stopped in steam but I can't open it again unless I restart the steam client. Same behavior for Half-Life 1.

**Comment by denisstrizhkin** [comment]
I can provide any additional info if needed.

**Comment by DeedleFake** [comment]
@EoceneMiacid

I've found that it _is_ possible to start a new game without rebooting, but it required killing the entire process tree of everything gamescope started and even then it seems to fail to fix the problem occasionally.

**Comment by HugLifeTiZ** [comment]
> > A little debugging show that it is stuck on
> > https://github.com/Plagman/gamescope/blob/2a09fbcb695102e9a990ed7cf1fdc350397a58f1/src/wlserver.cpp#L1211
> > Adding a timeout fix the issue, however I don't know why it is stuck there.
> 
> This seems like something that should be in a pull request.

A pull request was made, but @emersion felt that waiting was not a good fix. What is the problem with waiting? Specifically, what other things stand to break if this method is used? This is a legitimate question. Users are left out in the cold on a 7-month old issue that has a very trivial workaround, so it merits a response.

A *real fix* will definitely take some digging into all the interlocking parts, but there has been radio silence from gamescope contributors on this issue, so we don't know what--if anything--is being done to figure out what the *real* fix is. So why not merge a workaround if the workaround doesn't impact anything else? Sure, this creates technical debt, which is very real and shouldn't be ignored, but by leaving issues unaddressed in pursuit of real fixes, you create a different sort of debt that I can't really find a word for.

**Comment by yaomtc** [comment]
This also occurs with [xivlauncher](https://github.com/goatcorp/XIVLauncher.Core/). I use Gamescope to get around this issue: https://bugs.kde.org/show_bug.cgi?id=458233

**Comment by returnofblank** [comment]
I'm having the issue with CS2, can't say I've ever had the issue on other games, but I don't really use gamescope with other games.

**Comment by Joshua-Ashton** [comment]
Very strange, this used to work and no code has changed here. Gamescope sets itself up as a reaper so if it dies its children should too.

**Comment by yaomtc** [comment]
To be more specific, the XIVLauncher process ends just fine, it's specifically the gamescope-wl process that keeps going (and Steam doesn't know XIVLauncher has quit until I end that process manually)

**Comment by Joshua-Ashton** [comment]
Can you try with the latest stuff I pushed to master?

**Comment by shindaseishin** [comment]
@Joshua-Ashton I just pulled your changes and compiled. I have tested with a couple of games and gamescope now seems to quit properly when quitting the game.

**Comment by yaomtc** [comment]
Can't get it to work in Plasma, gives an error about something being unsupported. I'll get it in a minute. In GNOME there's a different issue.

```
$ XL_SECRET_PROVIDER=FILE gamescope -w 2560 -h 1440 -f -- %command% run --parent-expose-pids --parent-share-pids --parent-pid=1 --branch=stable --arch=x86_64 --command=xivlauncher dev.goats.xivlauncher
No CAP_SYS_NICE, falling back to regular-priority compute and threads.
Performance will be affected.
wlserver: [backend/headless/backend.c:68] Creating headless backend
ATTENTION: default value of option vk_khr_present_wait overridden by environment.
vulkan: selecting physical device 'AMD Radeon RX 5700 XT (RADV NAVI10)': queue family 1 (general queue family 0)
vulkan: physical device supports DRM format modifiers
vulkan: supported DRM formats for sampling usage:
vulkan:   AR24 (0x34325241)
vulkan:   XR24 (0x34325258)
vulkan:   AB24 (0x34324241)
vulkan:   XB24 (0x34324258)
vulkan:   RG16 (0x36314752)
vulkan:   NV12 (0x3231564E)
vulkan:   AB4H (0x48344241)
vulkan:   XB4H (0x48344258)
vulkan:   AB48 (0x38344241)
vulkan:   XB48 (0x38344258)
vulkan:   AB30 (0x30334241)
vulkan:   XB30 (0x30334258)
vulkan:   AR30 (0x30335241)
vulkan:   XR30 (0x30335258)
wlserver: Running compositor on wayland display 'gamescope-0'
wlserver: [backend/headless/backend.c:16] Starting headless backend
wlserver: [xwayland/sockets.c:63] Failed to bind socket @/tmp/.X11-unix/X0: Address already in use
wlserver: [xwayland/server.c:108] Starting Xwayland on :3
wlserver: [types/wlr_compositor.c:681] New wlr_surface 0x5650f2723df0 (res 0x5650f2749d30)
wlserver: [xwayland/server.c:273] Xserver is ready
pipewire: stream state changed: connecting
pipewire: stream state changed: paused
pipewire: stream available on node ID: 93
xwm: execvp failed: No such file or directory
gamescope: children shut down!
(EE) failed to read Wayland events: Broken pipe
```

**Comment by yaomtc** [comment]
Plasma: 

```
$ XL_SECRET_PROVIDER=FILE gamescope -w 2560 -h 1440 -f -- %command% run --parent-expose-pids --parent-share-pids --parent-pid=1 --branch=stable --arch=x86_64 --command=xivlauncher dev.goats.xivlauncher
No CAP_SYS_NICE, falling back to regular-priority compute and threads.
Performance will be affected.
Your Wayland compositor does NOT support wp_presentation/presentation-time which is required for VK_KHR_present_wait and VK_KHR_present_id which is needed for Gamescope to function.
Please update your compositor or complain to your compositor vendor for support.
```

**Comment by EoceneMiacid** [comment]
I just compiled the latest version from git as well, and I'm now experiencing several new problems.

I will update this comment as I understand better what is going on.

**Comment by awsms** [comment]
to add something to yaomtc comment, gamescope used to work a few months ago (toggling fullscreen will freeze my PC and other shenanigans). however it's now completely unusable (RTX 3600 mobile / Cinnamon DE): every single time I'll launch gamescope, I'll meet another bug then crash. and commit after commit, I'll meet even new bugs.
now, launching gamescope will simply invoke a black screen, and my PC will halt for ~5sec when I kill it.

```
gamescope -W 1920 -H 1080 -r 120 -- vkcube
No CAP_SYS_NICE, falling back to regular-priority compute and threads.
Performance will be affected.
wlserver: [backend/headless/backend.c:68] Creating headless backend
vulkan: selecting physical device 'NVIDIA GeForce RTX 3060 Laptop GPU': queue family 2 (general queue family 0)
vulkan: physical device supports DRM format modifiers
vulkan: vkGetPhysicalDeviceFormatProperties2 returned zero modifiers for DRM format 0x3231564E (VkResult: 0)
vulkan: vkGetPhysicalDeviceFormatProperties2 returned zero modifiers for DRM format 0x48344241 (VkResult: 0)
vulkan: vkGetPhysicalDeviceFormatProperties2 returned zero modifiers for DRM format 0x48344258 (VkResult: 0)
vulkan: vkGetPhysicalDeviceFormatProperties2 returned zero modifiers for DRM format 0x38344241 (VkResult: 0)
vulkan: vkGetPhysicalDeviceFormatProperties2 returned zero modifiers for DRM format 0x38344258 (VkResult: 0)
vulkan: supported DRM formats for sampling usage:
vulkan:   AR24 (0x34325241)
vulkan:   XR24 (0x34325258)
vulkan:   AB24 (0x34324241)
vulkan:   XB24 (0x34324258)
vulkan:   RG16 (0x36314752)
vulkan:   AB30 (0x30334241)
vulkan:   XB30 (0x30334258)
vulkan:   AR30 (0x30335241)
vulkan:   XR30 (0x30335258)
wlserver: Running compositor on wayland display 'gamescope-0'
wlserver: [backend/headless/backend.c:16] Starting headless backend
wlserver: [xwayland/server.c:108] Starting Xwayland on :1
wlserver: [types/wlr_compositor.c:681] New wlr_surface 0x55d2910818a0 (res 0x55d2910843d0)
wlserver: [xwayland/server.c:273] Xserver is ready
pipewire: stream state changed: connecting
pipewire: stream state changed: paused
pipewire: stream available on node ID: 69
Selected GPU 0: NVIDIA GeForce RTX 3060 Laptop GPU, type: DiscreteGpu
[Gamescope WSI] Creating Gamescope surface: xid: 0x400000
[Gamescope WSI] Atom of T was wrong type. Expected XCB_ATOM_CARDINAL.
wlserver: [types/wlr_compositor.c:681] New wlr_surface 0x55d290f44f70 (res 0x55d2910895a0)
[Gamescope WSI] Atom of T was wrong type. Expected XCB_ATOM_CARDINAL.
[Gamescope WSI] Made gamescope surface for xid: 0x400000
[Gamescope WSI] Surface state:
  steam app id:                  0
  window xid:                    0x400000
  wayland surface res id:        5
  layer client flags:            0x0
  server hdr output enabled:     false
  hdr formats exposed to client: false
wlserver: [types/wlr_compositor.c:681] New wlr_surface 0x55d29110aa30 (res 0x55d2910897e0)
xwm: got the same buffer committed twice, ignoring.
The XKEYBOARD keymap compiler (xkbcomp) reports:
> Warning:          Unsupported maximum keycode 708, clipping.
>                   X11 cannot support keycodes above 255.
Errors from xkbcomp are not fatal to the X server
[Gamescope WSI] Creating swapchain for xid: 0x400000 - minImageCount: 3 - format: VK_FORMAT_B8G8R8A8_UNORM - colorspace: VK_COLOR_SPACE_SRGB_NONLINEAR_KHR - flip: true
[Gamescope WSI] Created swapchain for xid: 0x400000 - imageCount: 3
```

I can still run it OK from the TTY though.

**Comment by Joshua-Ashton** [comment]
> gamescope -w 1920 -h 1080 -f -- wine game.exe - this now crashes my compositor outright, most of the time. Interestingly, when it doesn't, gamescope now terminates properly after quitting the game. The crash could be a Wayfire bug.

If your compositor is crashing, then yes, that's a compositor bug. (Unless you see a GPU hang in dmesg)

**Comment by Joshua-Ashton** [comment]
Also you should not be using `-e` if you are not launching Steam.

**Comment by awsms** [comment]
> > gamescope -w 1920 -h 1080 -f -- wine game.exe - this now crashes my compositor outright, most of the time. Interestingly, when it doesn't, gamescope now terminates properly after quitting the game. The crash could be a Wayfire bug.
> 
> If your compositor is crashing, then yes, that's a compositor bug. (Unless you see a GPU hang in dmesg)

do you know which compositors work fine with gamescope?

**Comment by yaomtc** [comment]
Oh yeah I forgot 3.12.x doesn't work for me: #919

**Comment by awsms** [comment]
> to add something to yaomtc comment, gamescope used to work a few months ago (toggling fullscreen will freeze my PC and other shenanigans). however it's now completely unusable (RTX 3600 mobile / Cinnamon DE): every single time I'll launch gamescope, I'll meet another bug then crash. and commit after commit, I'll meet even new bugs. now, launching gamescope will simply invoke a black screen, and my PC will halt for ~5sec when I kill it.
> 
> ```
> gamescope -W 1920 -H 1080 -r 120 -- vkcube
> No CAP_SYS_NICE, falling back to regular-priority compute and threads.
> Performance will be affected.
> wlserver: [backend/headless/backend.c:68] Creating headless backend
> vulkan: selecting physical device 'NVIDIA GeForce RTX 3060 Laptop GPU': queue family 2 (general queue family 0)
> vulkan: physical device supports DRM format modifiers
> vulkan: vkGetPhysicalDeviceFormatProperties2 returned zero modifiers for DRM format 0x3231564E (VkResult: 0)
> vulkan: vkGetPhysicalDeviceFormatProperties2 returned zero modifiers for DRM format 0x48344241 (VkResult: 0)
> vulkan: vkGetPhysicalDeviceFormatProperties2 returned zero modifiers for DRM format 0x48344258 (VkResult: 0)
> vulkan: vkGetPhysicalDeviceFormatProperties2 returned zero modifiers for DRM format 0x38344241 (VkResult: 0)
> vulkan: vkGetPhysicalDeviceFormatProperties2 returned zero modifiers for DRM format 0x38344258 (VkResult: 0)
> vulkan: supported DRM formats for sampling usage:
> vulkan:   AR24 (0x34325241)
> vulkan:   XR24 (0x34325258)
> vulkan:   AB24 (0x34324241)
> vulkan:   XB24 (0x34324258)
> vulkan:   RG16 (0x36314752)
> vulkan:   AB30 (0x30334241)
> vulkan:   XB30 (0x30334258)
> vulkan:   AR30 (0x30335241)
> vulkan:   XR30 (0x30335258)
> wlserver: Running compositor on wayland display 'gamescope-0'
> wlserver: [backend/headless/backend.c:16] Starting headless backend
> wlserver: [xwayland/server.c:108] Starting Xwayland on :1
> wlserver: [types/wlr_compositor.c:681] New wlr_surface 0x55d2910818a0 (res 0x55d2910843d0)
> wlserver: [xwayland/server.c:273] Xserver is ready
> pipewire: stream state changed: connecting
> pipewire: stream state changed: paused
> pipewire: stream available on node ID: 69
> Selected GPU 0: NVIDIA GeForce RTX 3060 Laptop GPU, type: DiscreteGpu
> [Gamescope WSI] Creating Gamescope surface: xid: 0x400000
> [Gamescope WSI] Atom of T was wrong type. Expected XCB_ATOM_CARDINAL.
> wlserver: [types/wlr_compositor.c:681] New wlr_surface 0x55d290f44f70 (res 0x55d2910895a0)
> [Gamescope WSI] Atom of T was wrong type. Expected XCB_ATOM_CARDINAL.
> [Gamescope WSI] Made gamescope surface for xid: 0x400000
> [Gamescope WSI] Surface state:
>   steam app id:                  0
>   window xid:                    0x400000
>   wayland surface res id:        5
>   layer client flags:            0x0
>   server hdr output enabled:     false
>   hdr formats exposed to client: false
> wlserver: [types/wlr_compositor.c:681] New wlr_surface 0x55d29110aa30 (res 0x55d2910897e0)
> xwm: got the same buffer committed twice, ignoring.
> The XKEYBOARD keymap compiler (xkbcomp) reports:
> > Warning:          Unsupported maximum keycode 708, clipping.
> >                   X11 cannot support keycodes above 255.
> Errors from xkbcomp are not fatal to the X server
> [Gamescope WSI] Creating swapchain for xid: 0x400000 - minImageCount: 3 - format: VK_FORMAT_B8G8R8A8_UNORM - colorspace: VK_COLOR_SPACE_SRGB_NONLINEAR_KHR - flip: true
> [Gamescope WSI] Created swapchain for xid: 0x400000 - imageCount: 3
> ```
> 
> I can still run it OK from the TTY though.

doesn't even work on TTY anymore since recent commits.

**Comment by EoceneMiacid** [comment]
Yeah, gamescope is now more broken than it's ever been.

**Comment by Joshua-Ashton** [comment]
Are you talking about the same issue where it's your compositor crashing? If so, I don't really think there is anything I can do about that. If you want to bisect it might be useful information at least.

> doesn't even work on TTY anymore since recent commits.

Can you give info on your setup? It's working fine for me.

**Comment by EoceneMiacid** [comment]
> Are you talking about the same issue where it's your compositor crashing? If so, I don't really think there is anything I can do about that. If you want to bisect it might be useful information at least.
> 

I'm working with the compositor developers to fix it. It is related to a 'bad xwayland window', so at the very least this helps us fix an edge case.

It is difficult to nail down what the precise cause is because we are testing it on three very similar setups (Arch Linux, AMD Radeon, exact same compositor revision) and we are all experiencing crashes, but under very different circumstances.

What also doesn't help is that the recent commit broke my gamescope integration in my frontend. There it doesn't make the compositor crash (despite using the same command line parameters), but rather the gamescope process just terminates immediately after launching with the 'gamescope: children shut down!' message. The game window then launches regardless.

So, very inconsistent behaviour. We are looking into the compositor crash situation at the moment.

**Comment by Joshua-Ashton** [comment]
If the main child shuts down then Gamescope will correctly now yes. Are you launching in some wrapper that immediately kills itself?

I would like to help but things are still very vague. If you can give repro steps that'd be appreciated.

**Comment by awsms** [comment]
> If the main child shuts down then Gamescope will correctly now yes.

This is not the case for me. After launching `gamescope -w 1920 -h 1080 -r 60 -- vkcube`, I can see the `gamescope: children shut down!`, but gamescope will never close by itself untill I manually kill the process. I'm running 3.12.7.r0.gd1eb70b. NVIDIA RTX 3600 mobile, i3, without a compositor.

**Comment by anthr76** [comment]
I'm also facing this issue on KDE 6

I have launch arguments that look like this

`gamemoderun gamescope -f --force-grab-cursor --mangoapp --prefer-vk-device 1002:744c  --rt -W 3840 -H 2160 -w 3840 -h 2160 -r 60 -- env MANGOHUD="1" %command%`

At the time of writing I have this up until this commit installed https://github.com/ValveSoftware/gamescope/commit/bc0cf62d813048a93ac8c82994340c166d863833

**Comment by DeedleFake** [comment]
Try running gamescope with `SDL_VIDEODRIVER=x11`. That seems to have fixed it the vast majority of the time for me, as well as a ton of other bugs.

**Comment by BearsPunch** [comment]
> Try running gamescope with `SDL_VIDEODRIVER=x11`. That seems to have fixed it the vast majority of the time for me, as well as a ton of other bugs.

Can confirm that SDL_VIDEODRIVER=x11 fix the issue. After that explorer.exe and other win processes closing after game closed

**Comment by superboo07** [comment]
> Try running gamescope with `SDL_VIDEODRIVER=x11`. That seems to have fixed it the vast majority of the time for me, as well as a ton of other bugs.

Game refuses to launch when using this as a launch option for gamescope.

**Comment by DeedleFake** [comment]
@superboo07

Because of various changes, the `SDL_VIDEODRIVER=x11` fix no longer works. Instead, use `env -u WAYLAND_DISPLAY gamescope ...`.

**Comment by unit73e** [comment]
I have the same issue with wine:
- gamescope 3.14.29
- wine-9.15.r0.g92374493 ( TkG Staging Esync Fsync )
- dxvk 2.4
- sway 1.9
- linux 6.10.3
- mesa 24.1.5 (RX 6600)

The process gets stuck if I run the following command and try to exit:

```
gamescope -- wine explorer
```

But it works fine if I run with the SDL backend:

```
gamescope --backend sdl -- wine explorer
```

With SDL I get a core dump though.

```
[1]    10904 IOT instruction (core dumped)  gamescope --backend sdl -- wine explorer
```

Better than not quitting. Not sure if that's normal because wine doesn't support Wayland AFAIK

**Comment by AlexFolland** [comment]
This happens to me as well.  I'm using KDE Plasma (Wayland).  Prepending `env -u WAYLAND_DISPLAY ` to my gamescope command does allow gamescope to quit, but leaves gamescopereaper running.  I have to periodically open ksysguard and send `KILL` signals to the `gamescopereaper` processes to get them to all quit.  Here's my gamescope command:

```Plaintext
env -u WAYLAND_DISPLAY gamescope --fullscreen --nested-width 1920 --nested-height 1080 --scaler fit --filter pixel -- wine '/run/media/alex/gaming/games/worms armageddon/WA.exe'
```

**Comment by Tatsh** [comment]
> This happens to me as well. I'm using KDE Plasma (Wayland). Prepending `env -u WAYLAND_DISPLAY ` to my gamescope command does allow gamescope to quit, but leaves gamescopereaper running. I have to periodically open ksysguard and send `KILL` signals to the `gamescopereaper` processes to get them to all quit. Here's my gamescope command:
> 
> ```
> env -u WAYLAND_DISPLAY gamescope --fullscreen --nested-width 1920 --nested-height 1080 --scaler fit --filter pixel -- wine '/run/media/alex/gaming/games/worms armageddon/WA.exe'
> ```

Basically the same and I have a script I do not like to use:

```
#!/usr/bin/env bash
getpids() {
    pidof gamescopereaper
    pidof gamescope
}
mapfile -t pids < <( getpids )
if (( ${#pids[@]} )); then
    kill -KILL "${pids[@]}"
fi
```

**Comment by unit73e** [comment]
@Tatsh an alternative is to use killwine: https://github.com/begin-theadventure/killwine

**Comment by fleurc** [comment]
@unit73e Thank you for referring to this, it at least solves the problem of gamescope being stuck until reboot for me. Though I'll subscribe to see when this issue is fixed and closed.

**Comment by unit73e** [comment]
@fleurc yes me too because it's slightly annoying. I did discover that games that don't exit correctly have this problem:
- Killing the window with `sigterm` can freeze the game and gamescope.
- Closing the window with alt+f4 or the close button may exit the game but not gamescope.
- Closing with a menu option usually exists the game and gamescope correctly.

Point being some games implement exit incorrectly. sigterm, closing the window or exit in game menu should all do the same thing, but that's not always the case.

I did look at gamescope code at the time and IIRC what it does is wait for the exit event. If that never happens it will wait forever. I could be misremembering and be completely wrong though. Not sure if it's worth looking at the code again because the latest commits might have already fixed the problem. That I could test for you guys by using gamescope-git in arch linux.

**Comment by foriequal0** [comment]
I use `systemd-run` to clean up child process at once.
It can be used to translate SIGTERM to SIGINT to child processes with `trap`

```shell
#!/usr/bin/env bash

# file: sigterm-to-sigint.sh

SCOPE="gamescope-$$"

handler() {
    systemctl --user kill --signal=SIGINT "$SCOPE.slice"
	systemctl --user stop "$SLICE.slice"
}
trap handler SIGTERM

systemd-run --user --scope --slice "$SCOPE" \
  --property=KillMode=mixed \
  --property=TimeoutStopSec=3 \
  -- "$@"
```

```shell
gamescope -- sigterm-to-sigint.sh wine explorer
```

**Comment by Tatsh** [comment]
> I use `systemd-run` to clean up child process at once. It can be used to translate SIGTERM to SIGINT to child processes with `trap`
> 
> #!/usr/bin/env bash
> 
> # file: sigterm-to-sigint.sh
> 
> SCOPE="gamescope-$$"
> 
> handler() {
>     systemctl --user kill --signal=SIGINT "$SCOPE.slice"
> }
> trap handler SIGTERM
> 
> systemd-run --user --scope --slice "$SCOPE" -- "$@"
> gamescope -- sigterm-to-sigint.sh wine explorer

Tried it with Sims 1 and it just doesn't exit cleanly.

**Comment by Letus252** [comment]
Is there any update on this issue? I’m playing CS2 with gamescope (+ mangoapp), and half of the time, when closing the game using the in-game console `quit` command, the entire gamescope window freezes, and I have to run `pkill -9 gamescope` (which still doesn’t immediately free up RAM).

**Comment by zlice** [comment]
is that even valid? void...return func?

```
    void KillProcessTree( std::vector<pid_t> nPids, int nSignal )
    {
        for ( pid_t nPid : nPids )
        {
            auto nChildPids = GetChildPids( nPid );
            KillProcess( nPid, nSignal );
            KillProcessTree( nChildPids, nSignal );
        }
    }

    void KillAllChildren( pid_t nParentPid, int nSignal )
    {
        std::vector<pid_t> nChildPids = GetChildPids( nParentPid );
        return KillProcessTree( nChildPids, nSignal );
    }
```

**Comment by DeedleFake** [comment]
Yep, it's valid: https://stackoverflow.com/a/2249281

**Comment by zlice** [comment]
if some people who use gamescope more could test that PR https://github.com/ValveSoftware/gamescope/pull/1908 that'd be great. i do not use gamescope a bunch so my test case is booting up solitare and making sure it dies

**Comment by SolarBakha** [comment]
> if some people who use gamescope more could test that PR [#1908](https://github.com/ValveSoftware/gamescope/pull/1908) that'd be great. i do not use gamescope a bunch so my test case is booting up solitare and making sure it dies

I can confirm that it does solve the problem. I don't use gamescope much so can't say I went that in-depth. But with the patch applied, in KDE Plasma on Wayland be it closing a game from it's menu/titlebar/alt-f4 it worked properly.

**Comment by Tatsh** [comment]
> if some people who use gamescope more could test that PR [#1908](https://github.com/ValveSoftware/gamescope/pull/1908) that'd be great. i do not use gamescope a bunch so my test case is booting up solitare and making sure it dies

That patch works for me. Tested Sims 1 which always hangs Gamescope and gamescopereaper after exiting. It exited without a problem.

**Comment by Inu90** [comment]
> if some people who use gamescope more could test that PR [#1908](https://github.com/ValveSoftware/gamescope/pull/1908) that'd be great. i do not use gamescope a bunch so my test case is booting up solitare and making sure it dies

Confirming patch also working for me. Tested on latest git,

**Comment by joguSD** [comment]
I ran into a similar issue as what's being seen here. I noticed that when run under gamescope when the game closed there was a child process that wasn't exiting:  `winedevice.exe`. If I killed that then gamescope would exit. Seems like there's a separate issue for this process hanging around and wrapping my wine call with the code in the linked comment fixed the issue for me:

https://github.com/ValveSoftware/gamescope/issues/1482#issuecomment-2562310285

**Comment by steviator** [comment]
Was this fix ever accepted? Having to reboot after every game for months is getting pretty old now.

Do I need to fork gamescope to get the fix integrated?

**Comment by awsms** [comment]
> Do I need to fork gamescope to get the fix integrated?

yes.

**Comment by K1NGSSTH** [comment]
I still have this problem so i created a very basic script to make sure gamescope closes properly, if anyone wants to use it or make it better here it is (i'm a very noob with scripts and no i did't use AI):

```
#!/bin/bash

while true ; do # This will ensure the script keps looping.
	while [[ 'pgrep ".exe" < 0' || 'pgrep "./" < 0' || 'pgrep "_linux" < 0' ]] ; do # This will see if there's any process with these set of characters closed.
		: # If it is closed it won't do nothing.
	done
	while [[ 'pgrep ".exe" > 0' || 'pgrep "./" > 0' || 'pgrep "_linux" > 0' ]] ; do # This will see if there's any process with these set of characters open.
		if [[ 'pgrep ".exe" > 0' || 'pgrep "./" > 0' || 'pgrep "_linux" > 0' ]] ; then # Same as above, this is to ensure it won't do anything while these apps are open.
			:
		else
   			kill -9 $(pgrep gamescope) # When you close them then this command will find any process with "gamescope" in it and will close it.
		fi
	done
	sleep 1 # A little timer to wait to go to the next loop.
done
```

Also i'm not sure it will work with every game, specially the native ones, i noticed that some native games always has the "./" in their names but Valve games like Half-Life don't but they name it with a "_linux" in front, don't know if is the standard for every game of theirs though.
